sdk/identity/identity/src/credentials/azureCredential.ts

```diff
@@ -0,0 +1,29 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+import { AccessToken } from './accessToken';
+import { TokenCredential, RequestOptionsBase } from '@azure/core-http'
+import { IdentityClient, IdentityClientOptions } from '../client/identityClient'
+
+export abstract class AzureCredential implements TokenCredential {
+  private cachedToken: AccessToken | null = null;
+  private refreshBufferMs: number;
+  protected identityClient: IdentityClient;
+
+  constructor(options?: IdentityClientOptions) {
+    options = options || IdentityClient.getDefaultOptions()
+    this.identityClient = new IdentityClient(options)
+    this.refreshBufferMs = options.refreshBufferMs
+  }
+
+  public async getToken(scopes: string | string[], requestOptions?: RequestOptionsBase): Promise<string | null> {
+    if (this.cachedToken && new Date(Date.now() + this.refreshBufferMs) < this.cachedToken.expiresOn) {
+      return this.cachedToken.token
+    }
+
+    this.cachedToken = await this.getTokenCore(scopes, requestOptions)
+    return this.cachedToken != null ? this.cachedToken.token : null;
+  }
+
+  protected abstract getTokenCore(scopes: string | string[], requestOptions?: RequestOptionsBase): Promise<AccessToken | null>;
```

Comment:
If I understood this correctly, then `getTokenCore` gets the token if there is no valid cached one available.

In that case, suffixing it with `core` seems .... off.
Is this the name that we chose to have in other languages as well?
Since it is not exposed to the user, am not too worried about the name, but for a second the word `core` threw me off.

Reply:
@schaabs do you have a link to the API overview `.md` file you prepared for the review meeting?  I can't remember if it was spec'ed to have this name.  At any rate, happy to change it to something clearer if it doesn't need to be consistent with the C# implementation.

Reply:
This API isn't public so it's not in the review doc.  In .NET this is called "core" because that's the convention for a protected method which provides an implementation of a public method to the base class, but this doesn't need to be consistent across languages.

---
In reply to: [290941176](https://github.com/Azure/azure-sdk-for-js/pull/3449#discussion_r290941176) [](ancestors = 290941176)

Reply:
Thanks Scott!  Perhaps I'll rename it to `getAccessToken` since the method returns an `AccessToken` object.